File: le-transform/src/main/java/com/latticeengines/transform/v2_0_25/functions/StdVisidbDsIndustryGroup.java
package com.latticeengines.transform.v2_0_25.functions;

import java.util.Map;
import java.util.regex.Pattern;

import org.apache.commons.lang3.StringUtils;

import com.latticeengines.transform.exposed.RealTimeTransform;
import com.latticeengines.transform.exposed.metadata.ApprovedUsage;
import com.latticeengines.transform.exposed.metadata.Category;
import com.latticeengines.transform.exposed.metadata.FundamentalType;
import com.latticeengines.transform.exposed.metadata.StatisticalType;
import com.latticeengines.transform.exposed.metadata.Tag;
import com.latticeengines.transform.exposed.metadata.TransformMetadata;

public class StdVisidbDsIndustryGroup implements RealTimeTransform {

    private static final long serialVersionUID = -5157087545940247272L;

    public StdVisidbDsIndustryGroup() {
    }

    public StdVisidbDsIndustryGroup(String modelPath) {
    }

    public Object transform(Map<String, Object> arguments, Map<String, Object> record) {
        String column = (String) arguments.get("column");
        String s = column == null ? null : String.valueOf(record.get(column));

        if ("null".equals(s))
            return null;

        return calculateStdVisidbDsIndustryGroup(s);
    }

    public static String calculateStdVisidbDsIndustryGroup(String industryGroup) {
        String stdGroup = null;
        if (StringUtils.isNotBlank(industryGroup)) {
            industryGroup = industryGroup.trim().toLowerCase();
            if (Pattern.matches("(.*?\\b)credit(.*?\\b)|(.*?\\b)financial(.*?\\b)|(.*?\\b)bank(.*?\\b)", industryGroup))
                stdGroup =  "Finance";
            else if (Pattern.matches("(.*)tech(.*)|(.*?\\b)tele(.*?\\b)", industryGroup))
                stdGroup =  "Tech";
            else if (Pattern.matches("(.*?\\b)health(.*?\\b)|(.*?\\b)medical(.*?\\b)|(.*?\\b)pharm(.*?\\b)", industryGroup))
                stdGroup =  "Health Care";
            else if (Pattern.matches("(.*?\\b)real(.*?\\b)|(.*?\\b)property(.*?\\b)", industryGroup))
                stdGroup =  "Real Estate/Property Mgmt";
            else if (Pattern.matches("(.*?\\b)staff(.*?\\b)|(.*?\\b)hr(.*?\\b)", industryGroup))
                stdGroup =  "HR/Staffing";
            else if (Pattern.matches("(.*?\\b)services(.*?\\b)|(.*?\\b)consulting(.*?\\b)", industryGroup))
                stdGroup =  "Business Service";
            else if (Pattern.matches("(.*?\\b)education(.*?\\b)", industryGroup))
                stdGroup =  "Education";
            else if (Pattern.matches("(.*?\\b)equipment(.*?\\b)", industryGroup))
                stdGroup =  "Equipment";
            else if (Pattern.matches("(.*?\\b)util(.*?\\b)", industryGroup))
                stdGroup =  "Utilities";
            else if (Pattern.matches("(.*?\\b)retail(.*?\\b)", industryGroup))
                stdGroup =  "Retail";
            else if (Pattern.matches("(.*?\\b)transport(.*?\\b)", industryGroup))
                stdGroup =  "Transportation";
            else if (Pattern.matches("(.*?\\b)account(.*?\\b)|(.*?\\b)legal(.*?\\b)", industryGroup))
                stdGroup =  "Accounting/Legal";
            else if (Pattern.matches("(.*?\\b)construction(.*?\\b)", industryGroup))
                stdGroup =  "Construction";
            else if (Pattern.matches("(.*?\\b)profit(.*?\\b)", industryGroup))
                stdGroup =  "Non-Profit";
            else if (Pattern.matches("(.*?\\b)insurance(.*?\\b)", industryGroup))
                stdGroup =  "Insurance";
            else if (Pattern.matches("(.*?\\b)manufact(.*?\\b)", industryGroup))
                stdGroup =  "Manufacturing";
            else
                stdGroup =  "Other";
        }
        return stdGroup;
    }

    @Override
    public TransformMetadata getMetadata() {
        TransformMetadata metadata = new TransformMetadata();
        metadata.setApprovedUsage(ApprovedUsage.MODEL_ALLINSIGHTS);
        metadata.setCategory(Category.LEAD_INFORMATION);
        metadata.setFundamentalType(FundamentalType.ALPHA);
        metadata.setStatisticalType(StatisticalType.NOMINAL);
        metadata.setTags(Tag.INTERNAL_TRANSFORM);
        metadata.setDescription("Rollup of Industry field from Marketing Automation");
        metadata.setDisplayName("Industry Rollup");
        return metadata;
    }
}

File: le-serviceflows/core/src/main/resources/com/latticeengines/serviceflows/core/transforms/std_visidb_ds_industry_group.py

#
# $LastChangedBy$
# $LastChangedDate$
# $Rev$
#

import re

def metadata():
    return {  'ApprovedUsage'   : 'ModelAndAllInsights'
            , 'Description'     : 'Rollup of Industry field from Marketing Automation'
            , 'DisplayName'     : 'Industry Rollup'
            , 'Category'        : 'Lead Information'
            , 'FundamentalType' : 'alpha'
            , 'StatisticalType' : 'nominal'
            , 'Tags'            : 'InternalTransform' }


def std_visidb_ds_industry_group(n):
    if n is None or n == '':
        return None

    n = n.lower()

    if re.search('credit|financial|bank', n):
        return 'Finance'
    elif re.search('tech|tele', n):
        return 'Tech'
    elif re.search('health|medical|pharm', n):
        return 'Health Care'
    elif re.search('real|property', n):
        return 'Real Estate/Property Mgmt'
    elif re.search('staff|hr', n):
        return 'HR/Staffing'
    elif re.search('services|consulting', n):
        return 'Business Service'
    elif re.search('education', n):
        return 'Education'
    elif re.search('equipment', n):
        return 'Equipment'
    elif re.search('util', n):
        return 'Utilities'
    elif re.search('retail', n):
        return 'Retail'
    elif re.search('transport', n):
        return 'Transportation'
    elif re.search('account|legal', n):
        return 'Accounting/Legal'
    elif re.search('construction', n):
        return 'Construction'
    elif re.search('profit', n):
        return 'Non-Profit'
    elif re.search('insurance', n):
        return 'Insurance'
    elif re.search('manufact', n):
        return 'Manufacturing'

    return 'Other'


def transform(args, record):
    column = args["column"]
    value = record[column]
    return std_visidb_ds_industry_group(value)
